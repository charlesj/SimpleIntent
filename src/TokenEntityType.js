import natural, { NGrams } from 'natural'

export const TokenEntityType = (name, values) => {
  const tokenizer = new natural.WordTokenizer()

  const tokenizedValues = Object.entries(values).reduce((acc, [entity, synonyms]) => {
    synonyms.forEach(syn => {
      acc.push({ entity, synonymTokens: tokenizer.tokenize(syn) })
    })
    return acc
  }, [])

  const match = (tokens) => {
    const matches = tokenizedValues.map(({ entity, synonymTokens }) => {
      if (synonymTokens.length === 1) {
        const matches = matchSingleTokens(tokens, synonymTokens[0])
        if (matches.length > 0) {
          return matches.map(m => ({ entity, ...m }))
        }
      } else {
        return matchPhrase(tokens, synonymTokens).map(m => ({ entity, ...m }))
      }
    }).filter(m => m).reduce((acc, matches) => acc.concat(matches), [])

    return matches
  }

  return {
    entityType: name,
    match,
  }
}

const hammingMatch = (source, target) => {
  const distance = natural.HammingDistance(source, target, true)
  return distance >= 0 && distance < 2
}

const matchPhrase = (tokens, synonymTokens) => {
  const grammified = NGrams.ngrams(tokens.join(' '), synonymTokens.length)
  const matched = grammified.filter((ngrams) => {
    return ngrams.every((token, index) => hammingMatch(token, synonymTokens[index]))
  })
  return matched.map(m => ({
    value: m,
    position: tokens.indexOf(m[0])
  }))
}



const matchSingleTokens = (tokens, synonym) => {
  const matched = tokens.map((token, index) => {
    if (hammingMatch(synonym, token)) {
      return { token, index }
    }
  }).filter(e => e)

  return matched.map(m => ({
    value: m.token,
    position: m.index,
  }))
}
