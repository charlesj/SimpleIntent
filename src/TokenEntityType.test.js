import { TokenEntityType } from './TokenEntityType'

it('can match a simple entity', () => {
  const entity = TokenEntityType('simple', { simple: ['simple'] })
  expect(entity.match(['simple'])).toMatchSnapshot()
})

it('does not match a simple entity against something else', () => {
  const entity = TokenEntityType('simple', { simple: ['simple'] })
  expect(entity.match(['wat'])).toMatchSnapshot()
})

it('can match phrases', () => {
  const entity = TokenEntityType('phrase', { simple: ['multiple words'] })
  expect(entity.match(['here', 'are', 'multiple', 'words'])).toMatchSnapshot()
})
