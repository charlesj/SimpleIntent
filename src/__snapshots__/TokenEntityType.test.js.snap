// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can match a simple entity 1`] = `
Array [
  Object {
    "entity": "simple",
    "position": 0,
    "value": "simple",
  },
]
`;

exports[`can match phrases 1`] = `
Array [
  Object {
    "entity": "simple",
    "position": 2,
    "value": Array [
      "multiple",
      "words",
    ],
  },
]
`;

exports[`does not match a simple entity against something else 1`] = `Array []`;
